### PR TITLE
NO-ISSUE WS7: Add core/format — output formatting

### LINKED ISSUE #9: WS7: core/format — Output Formatting
## WS7: core/format — Output Formatting

**Phase**: 3 (depends on WS1)
**Branch**: `ws7/core-format`
**Depends on**: #3
**Tracking**: #2

### Purpose

Implement human-readable and JSON output formatters in the MoonBit core for consistent output across all hosts.

### Tasks

- [ ] Write `core/format/README.md`
- [ ] Write test cases matching exact output from root README:
  - [ ] Human format: WARN/FAIL prefixes, indented messages, summary line
  - [ ] JSON format: exact structure from README spec
  - [ ] Edge cases: zero findings, single finding, many findings
- [ ] Implement human formatter: `ScanResult -> String`
- [ ] Implement JSON formatter: `ScanResult -> String`

### Design Choice

Formatting lives in the core (MoonBit) so both WASM and JS targets produce identical output across all hosts.

### PR DESCRIPTION
**Pull request by maruloop**
## Summary

- Implements `format_human(ScanResult) -> String` and `format_json(ScanResult) -> String`
- Human format: `WARN`/`FAIL` labels, optional context in parens, message, `Tip:` suggestion, summary line with correct pluralization
- JSON format: target as `"owner/repo@ref"` string, `ref` field, lowercase level values, optional fields omitted when absent
- 11 tests covering: zero findings, single warning/failure, context, suggestion, multiple findings, pluralization, JSON structure, omitted fields

## Test plan

- [x] `moon test` — 100 tests pass (all packages)
- [x] `moon info && moon fmt` — clean

Closes #9
Tracking: #2

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: core/format/README.mbt.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    "warnings": 1
+  }
+}
+```
+
+Rules:
+- `target` is `"owner/repo@ref"` string (not a nested object)
+- `ref` is the `git_ref` from `ScanTarget`
+- `level` values are lowercase: `"warn"` or `"fail"`
+- Omit `target`, `context`, and `suggestion` fields from findings when absent (`None`)
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/README.mbt.md]: Fixed — removed the spurious "use " prefix from the Tip example in README so it matches the actual formatter output. Committed in 960de13.

File: core/format/README.mbt.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+pub fn format_human(result : @papion.ScanResult) -> String
+pub fn format_json(result : @papion.ScanResult) -> String
+```
+
+## Human format
+
+```
+  WARN  actions/checkout@v4
+        Referenced by tag, not pinned to a SHA.
+        Tip: use actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/README.mbt.md]: Fixed — target is a required String field on Finding, so the JSON rules now only mention omitting context and suggestion when None. Committed in 960de13.

File: core/format/format.mbt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    plural_form
+  }
+}
+
+///|
+/// Format a ScanResult as human-readable text.
+///
+/// Each finding shows WARN/FAIL label, target (with optional context), message,
+/// and optional suggestion. Findings separated by blank lines.
+/// Summary line at the end: "N findings  (F failure(s), W warning(s))".
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/format.mbt]: Fixed in 50150be — docstring now reads: "Summary line uses singular/plural: 'N finding/findings  (F failure/failures, W warning/warnings)'".

File: core/format/README.mbt.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+```
+  WARN  actions/checkout@v4
+        Referenced by tag, not pinned to a SHA.
+        Tip: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
+
+  FAIL  actions/github-script@v7 (used in composite step "setup")
+        Not pinned to a SHA.
+
+2 findings  (1 failure, 1 warning)
+```
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/README.mbt.md:23]: Fixed in 50150be — root README.md also updated to match (committed locally, pending a direct-to-main PR).

File: core/format/README.mbt.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+Rules:
+- Each finding starts with 2-space indent, level label (`WARN` or `FAIL`), 2 spaces, then the target
+- If the finding has a `context`, append ` (context)` to the target line
+- Message on the next line, indented 8 spaces
+- If the finding has a `suggestion`, add `Tip: suggestion` on the line after the message (8-space indent)
+- Blank line between findings
+- Summary line at the end: `N findings  (F failure(s), W warning(s))`
+  - Use singular "failure"/"warning" when count is 1
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/README.mbt.md]: Fixed in 50150be — rule now explicitly covers "finding": use singular "finding"/"failure"/"warning" when the corresponding count is 1.

File: core/format/format_test.mbt [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+  let output = format_json(result)
+  let parsed = @json.parse(output)
+  match parsed {
+    Object(obj) =>
+      match obj.get("findings") {
+        Some(Array(arr)) => {
+          assert_eq(get_str(arr[0], "level"), Some("fail"))
+          assert_eq(get_str(arr[0], "context"), None)
+          assert_eq(get_str(arr[0], "suggestion"), None)
+        }
```
[the remark this answers is not in the document]
Comment (maruloop) [reply core/format/format_test.mbt]: Fixed in 50150be — test now directly checks `finding_obj.get("context") == None` and `finding_obj.get("suggestion") == None` to confirm the keys are absent, not just missing as strings.
